### PR TITLE
Beginning C++23: Chapter 14

### PR DESCRIPTION
**Pull request by Peter-VanWeert**

### CODE REVIEW

File: Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp [hunk 1]
```diff
@@ -1,7 +1,7 @@
 // Defining and using a derived class
-import <iostream>;
-import box;                 // For the Box class
-import carton;              // For the Carton class
+import <print>;
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp:2]: The text in this chapter uses "import std;", but the code not yet.
On purpose?
I won't repeat for the other files.

File: Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp [hunk 2]
```diff
@@ -1,7 +1,7 @@
...
+import box;
+import carton;
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp:4]: You removed the comments for the box and carton imports. On purpose?

File: Examples/Modules/Chapter 14/Ex14_02/Carton.cppm [hunk 3]
```diff
@@ -3,28 +3,29 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
 
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; }  
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_02/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_02/Carton.cppm [hunk 4]
```diff
@@ -3,28 +3,29 @@ export module carton;
...
+
 /*
   // This constructor won't compile!
   Carton::Carton(double l, double w, double h, std::string_view material)
     : m_length{ l }, m_width{ w }, m_height{ h }, m_material{ material }
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; }
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_02/Carton.cppm:28]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_02/Carton.cppm [hunk 5]
```diff
@@ -34,7 +35,7 @@ public:
     m_length = l;  // These should normally be initialized in a base class constructor...
     m_width = w;
     m_height = h;
-    std::cout << "Carton(double,double,double,string_view) called.\n";
+    std::println("Carton(double, double, double, string_view) called.)");
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_02/Carton.cppm:38]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_03/Carton.cppm [hunk 6]
```diff
@@ -3,31 +3,31 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
 
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; }  
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_03/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_04/Carton.cppm [hunk 7]
```diff
@@ -3,20 +3,23 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
+
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; }  
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_04/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_04A/Carton.cppm [hunk 8]
```diff
@@ -3,20 +3,23 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
+
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; }  
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_04A/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_04B/Carton.cppm [hunk 9]
```diff
@@ -3,20 +3,23 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
+
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; }  
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_04B/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_04C/Carton.cppm [hunk 10]
```diff
@@ -3,20 +3,23 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
+
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; }  
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_04C/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_05/Carton.cppm [hunk 11]
```diff
@@ -14,7 +14,7 @@ public:
   Carton() = default;   // Required by Visual C++
   Carton(double length, double width, double height, std::string_view mat)
      : Box{length, width, height}, m_material{mat}
-     { std::cout << "Carton(double,double,double,string_view) called.\n"; }
+     { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_05/Carton.cppm:17]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_07/Carton.cppm [hunk 12]
```diff
@@ -3,30 +3,30 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
 
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; } 
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_07/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_07/Carton.cppm [hunk 13]
```diff
@@ -3,30 +3,30 @@ export module carton;
...
 
   // One new constructor
   Carton(double l, double w, double h, std::string_view m, double density, double thickness)
     : Carton{l, w, h, m} 
   {
     m_thickness = thickness; m_density = density;
-    std::cout << "Carton(double,double,double,string_view,double,double) called.\n";
+    std::println("Carton(double, double, double, string_view, double, double) called.)");
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_07/Carton.cppm:29]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_07/Ex14_07.cpp [hunk 14]
```diff
@@ -1,19 +1,18 @@
...
-  std::cout << "cornflakes volume is " << cornflakes.volume() << std::endl
-            << "cornflakes weight is " << cornflakes.getWeight() << std::endl;
+  std::println("cornflakes volume is {:.3}", cornflakes.volume());
+  std::println("cornflakes weight is {:.3}", cornflakes.getWeight());
 /*
   // Here is one way to make this work (see Ex14_07A and B for alternatives)
-  std::cout << "cornflakes volume is " << cornflakes.Carton::volume() << std::endl
-          << "cornflakes weight is " << cornflakes.FoodContainer::getWeight() 
-          << std::endl;
+  std::println("cornflakes volume is {}", cornflakes.Carton::volume());
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_07/Ex14_07.cpp:15]: Shouldn't you specify {:.3} here as well?

File: Examples/Modules/Chapter 14/Ex14_07A/Carton.cppm [hunk 15]
```diff
@@ -3,30 +3,30 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
 
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; } 
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_07A/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_07A/Carton.cppm [hunk 16]
```diff
@@ -3,30 +3,30 @@ export module carton;
...
 
   // One new constructor
   Carton(double l, double w, double h, std::string_view m, double density, double thickness)
     : Carton{l, w, h, m} 
   {
     m_thickness = thickness; m_density = density;
-    std::cout << "Carton(double,double,double,string_view,double,double) called.\n";
+    std::println("Carton(double, double, double, string_view, double, double) called.)");
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_07A/Carton.cppm:29]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_07B/Carton.cppm [hunk 17]
```diff
@@ -3,30 +3,30 @@ export module carton;
...
+  { std::println("Carton(string_view) called."); }
 
   Carton(double side, std::string_view material) : Box{side}, m_material{material}
-  { std::cout << "Carton(double,string_view) called.\n"; }
+  { std::println("Carton(double,string_view) called."); }
 
   Carton(double l, double w, double h, std::string_view material)
     : Box{l, w, h}, m_material{material}
-    { std::cout << "Carton(double,double,double,string_view) called.\n"; } 
+    { std::println("Carton(double, double, double, string_view) called.)"); }
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_07B/Carton.cppm:22]: Remove spurious ) before last double quote.

File: Examples/Modules/Chapter 14/Ex14_07B/Carton.cppm [hunk 18]
```diff
@@ -3,30 +3,30 @@ export module carton;
...
 
   // One new constructor
   Carton(double l, double w, double h, std::string_view m, double density, double thickness)
     : Carton{l, w, h, m} 
   {
     m_thickness = thickness; m_density = density;
-    std::cout << "Carton(double,double,double,string_view,double,double) called.\n";
+    std::println("Carton(double, double, double, string_view, double, double) called.)");
```
Comment (MarcGregoire) [inline Examples/Modules/Chapter 14/Ex14_07B/Carton.cppm:29]: Remove spurious ) before last double quote.

**Review by MarcGregoire — COMMENTED** [review]
Done with review.

File: Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp (hunk 1 above)
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp:2]: Yes. For now, the code in this repo is such that it compiles, at least already with the latest Visual Studio (with, of course, a custom `print` header that forwards to `std::format()`. We'll see how compiler support evolves to know what I'll do with `import std;` for the online code samples.

File: Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp (hunk 2 above)
Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp:4]: Sure. We're defining classes and modules a few chapters already in Chapter 14. It's a bit overly obvious what an `import carton;` means, no?

Comment (Peter-VanWeert) [reply Examples/Modules/Chapter 14/Ex14_01/Ex14_01.cpp:4]: Changed my mind: I did now in fact change all standard library `import` declarations / `#include` directives to `import std;`, using a custom `std` module full of `export` `import`s to make it compile (thanks for that idea, btw!).

**Comment by Peter-VanWeert** [comment]
Missed 23 std::couts as well in the examples of Chapter 14 it seems...

**Comment by MarcGregoire** [comment]
> Missed 23 std::couts as well in the examples of Chapter 14 it seems...

Anything you want me to review?

**Pull request merged**